src/Tools/ExternalAccess/LiveShare/Tagger/SyntacticClassificationModeSelector.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Classification;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Tagger
+{
+    /// <summary>
+    /// this is almost straight copy from typescript for syntatic LSP experiement.
+    /// we won't attemp to change code to follow Roslyn style until we have result of the experiement
+    /// </summary>
+    internal sealed class SyntacticClassificationModeSelector
+    {
+        private readonly AbstractLspClientServiceFactory _lspClientServiceFactory;
+        private readonly ITextBuffer _buffer;
+
+        private int _modeCache = -1;
+
+        public SyntacticClassificationModeSelector(AbstractLspClientServiceFactory lspClientServiceFactory, ITextBuffer buffer)
+        {
+            _lspClientServiceFactory = lspClientServiceFactory;
+            _buffer = buffer;
+        }
+
+        public static SyntacticClassificationModeSelector GetModeSelector(AbstractLspClientServiceFactory lspClientServiceFactory, ITextBuffer buffer)
+        {
+            return new SyntacticClassificationModeSelector(lspClientServiceFactory, buffer);
+        }
+
+        public SyntacticClassificationMode GetMode()
+        {
+            if (_modeCache != -1)
+            {
+                return (SyntacticClassificationMode)_modeCache;
+            }
+
+            if (!Workspace.TryGetWorkspace(_buffer.AsTextContainer(), out var workspace))
+            {
+                return SyntacticClassificationMode.TextMate;
+            }
+
+            var experimentService = workspace.Services.GetService<IExperimentationService>();
+            if (experimentService == null)
+            {
+                return SyntacticClassificationMode.TextMate;
+            }
+
+            // this is most likely already ran, so it will be no-op
+            RoslynSyntaxClassificationService.EnsureInitializationAsync(_lspClientServiceFactory, CancellationToken.None).Wait(CancellationToken.None);
```

Comment:
any issues with just waiting here?  Idt ensure initialized should ever need to switch to the UI thread but I can't be 100% sure liveshare wouldn't

Reply:
it would be nice if I dont need to do this at all.